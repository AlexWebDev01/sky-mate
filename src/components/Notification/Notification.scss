.notification {
  display: flex;
  align-items: center;
  width: auto;
  height: auto;
  position: fixed;
  top: 130px;
  left: -220px;
  background-color: white;
  padding: 20px 40px 20px 15px;
  border-radius: 15px;
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
  transition: left 0.6s ease-in-out;
  z-index: 20;

  &.visible {
    left: 50px;
  }

  .text {
    color: var(--red);
    font-family: Inter, sans-serif;
    font-size: 16px;
    margin: 0;
  }

  .close-button {
    position: absolute;
    top: 10px;
    right: 10px;
    cursor: pointer;
    color: black;
    background-color: transparent;
  }
}
