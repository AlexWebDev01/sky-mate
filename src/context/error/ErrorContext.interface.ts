import { ReactNode } from 'react';

export interface ErrorContext {
  showNotification: (message: string) => void;
}

export interface ErrorProviderProps {
  children: ReactNode;
}
