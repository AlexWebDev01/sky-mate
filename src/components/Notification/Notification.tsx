import { useEffect, useRef, useState } from 'react';
import { useGlobalContext } from '@context/global/GlobalContext';

import './Notification.scss';

const Notification = () => {
  const { error, setError } = useGlobalContext();
  const [isVisible, setIsVisible] = useState<boolean>(false);
  const refreshTimeout = useRef<NodeJS.Timeout | null>(null);

  useEffect(() => {
    if (error) {
      setIsVisible(true);

      if (refreshTimeout.current) {
        clearTimeout(refreshTimeout.current);
      }

      refreshTimeout.current = setTimeout(() => {
        setIsVisible(false);
        setTimeout(() => {
          setError('');
        }, 600);
      }, 5000);
    }
  }, [error, setError]);

  const handleClose = () => {
    setIsVisible(false);
    setTimeout(() => {
      setError('');
    }, 600);
  };

  return (
    <div
      className={isVisible ? 'visible notification' : 'notification'}
      data-testid='notification'
    >
      <p className='text'>{error}</p>
      <button
        className='close-button'
        onClick={handleClose}
        data-testid='close-button'
      >
        X
      </button>
    </div>
  );
};

export default Notification;
