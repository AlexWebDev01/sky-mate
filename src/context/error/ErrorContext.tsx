import {
  useState,
  createContext,
  useContext,
  useMemo,
  FC,
  useEffect,
  useCallback,
} from 'react';
import {
  ErrorContext as ErrorContextInterface,
  ErrorProviderProps,
} from './ErrorContext.interface';
import Notification from '@components/Notification/Notification';

const ErrorContext = createContext<ErrorContextInterface | null>(null);

export const useErrorContext = () => {
  const context = useContext(ErrorContext);

  if (!context) {
    throw new Error('Error context should be used in ErrorProvider!');
  }

  return context;
};

export const ErrorProvider: FC<ErrorProviderProps> = ({ children }) => {
  const [error, setError] = useState<string>('');
  const [isVisible, setIsVisible] = useState<boolean>(false);

  const closeNotification = () => {
    setIsVisible(false);
  };

  const showNotification = useCallback((message: string) => {
    setError(message);
    setIsVisible(true);
  }, []);

  const sharedState = useMemo<ErrorContextInterface>(
    () => ({
      showNotification,
    }),
    [showNotification]
  );

  useEffect(() => {
    let timeout: NodeJS.Timeout;
    let timeout2: NodeJS.Timeout;

    if (isVisible) {
      timeout = setTimeout(() => {
        setIsVisible(false);
      }, 5000);
    }

    if (!isVisible && error) {
      timeout2 = setTimeout(() => {
        setError('');
      }, 600);
    }

    return () => {
      clearTimeout(timeout);
      clearTimeout(timeout2);
    };
  }, [isVisible]);

  return (
    <ErrorContext.Provider value={sharedState}>
      {children}
      {
        <Notification
          message={error}
          onClose={closeNotification}
          isVisible={isVisible}
        />
      }
    </ErrorContext.Provider>
  );
};
