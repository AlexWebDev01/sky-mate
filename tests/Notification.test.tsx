import React from 'react';
import '@testing-library/jest-dom';
import { render, screen, fireEvent, act } from '@testing-library/react';
import Notification from '@components/Notification/Notification';
import { ErrorProvider, useErrorContext } from '@context/error/ErrorContext';

describe('Notification Component', () => {
  test('renders the notification message', () => {
    render(
      <Notification
        message='Test message'
        onClose={() => {}}
        isVisible={true}
      />
    );

    const messageElement = screen.getByText(/Test message/i);
    expect(messageElement).toBeInTheDocument();
  });

  test('calls onClose when close button is clicked', () => {
    const onCloseMock = jest.fn();
    render(
      <Notification
        message='Test message'
        onClose={onCloseMock}
        isVisible={true}
      />
    );

    const closeButton = screen.getByTestId('close-button');
    fireEvent.click(closeButton);

    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  test('has the correct class when isVisible is true', () => {
    render(
      <Notification
        message='Test message'
        onClose={() => {}}
        isVisible={true}
      />
    );

    const notificationElement = screen.getByTestId('notification');
    expect(notificationElement).toHaveClass('visible notification');
  });

  test('has the correct class when isVisible is false', () => {
    render(
      <Notification
        message='Test message'
        onClose={() => {}}
        isVisible={false}
      />
    );

    const notificationElement = screen.getByTestId('notification');
    expect(notificationElement).toHaveClass('notification');
    expect(notificationElement).not.toHaveClass('visible');
  });
});

const TestComponent: React.FC = () => {
  const { showNotification } = useErrorContext();

  React.useEffect(() => {
    showNotification('Test error message');
  }, [showNotification]);

  return null;
};

describe('ErrorContext and Notification Component', () => {
  jest.useFakeTimers();

  test('shows the notification when the error message is setup', () => {
    render(
      <ErrorProvider>
        <TestComponent />
      </ErrorProvider>
    );

    expect(screen.getByText('Test error message')).toBeInTheDocument();
    expect(screen.getByTestId('notification')).toHaveClass(
      'visible notification'
    );
  });

  test('hides the notification after 5 seconds', () => {
    render(
      <ErrorProvider>
        <TestComponent />
      </ErrorProvider>
    );

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(screen.getByTestId('notification')).not.toHaveClass('visible');
  });

  test('resets the error message after hiding the notification', () => {
    render(
      <ErrorProvider>
        <TestComponent />
      </ErrorProvider>
    );

    act(() => {
      jest.advanceTimersByTime(5600);
    });

    act(() => {
      jest.advanceTimersByTime(600);
    });

    expect(screen.getByTestId('notification')).not.toHaveTextContent(
      'Test error message'
    );
  });
});
