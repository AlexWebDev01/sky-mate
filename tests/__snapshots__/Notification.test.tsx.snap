// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorContext and Notification Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="visible notification"
    data-testid="notification"
  >
    <p
      class="text"
    >
      Test error message
    </p>
    <button
      class="close-button"
      data-testid="close-button"
    >
      X
    </button>
  </div>
</DocumentFragment>
`;
